packages/db-kit/src/utils/fetchExternal.ts

```diff
@@ -0,0 +1,192 @@
+const { Compile } = require("@truffle/compile-solidity");
+import type TruffleConfig from "@truffle/config";
+import type { WorkflowCompileResult } from "@truffle/compile-common";
+import * as Common from "@truffle/compile-common";
+import {
+  Fetcher,
+  FetcherConstructor,
+  InvalidNetworkError,
+  default as Fetchers
+} from "@truffle/source-fetcher";
+import type { Project } from "@truffle/db";
+
+export interface FetchExternalOptions {
+  config: TruffleConfig;
+  project: Project.ConnectedProject;
+  address: string;
+}
+
+export async function fetchExternal({
+  config,
+  project,
+  address
+}: FetchExternalOptions): Promise<void> {
+  const { contractName, result } = await fetchAndCompile({ config, address });
+
+  const { contracts } = await project.loadCompile({ result });
+
+  const contract = await findContract({
+    contracts,
+    contractName
+  });
+
+  await project.assignNames({
+    assignments: {
+      contracts: [contract.db.contract]
+    }
+  });
+
+  const { network } = await project.loadMigrate({
+    network: { name: config.network },
+    artifacts: [
+      {
+        ...Common.Shims.NewToLegacy.forContract(contract),
+        networks: {
+          [config.network_id]: {
+            address
+          }
+        }
+      }
+    ]
+  });
+
+  await project.assignNames({
+    assignments: {
+      networks: [network]
+    }
+  });
+}
+
+/**
+ * Accepts a TruffleConfig and returns a priority-sorted list of fetchers based
+ * on provided configuration and whether available fetchers support the current
+ * blockchain network.
+ * @dev HACK this currently only supports EtherscanFetcher
+ */
+export async function forTruffleConfig(options: {
```

Comment:
Oh, factoring this out into its own function is neat.  Maybe this should go somewhere more accessible and then the debugger could use this?  (Of course, the restriction to Etherscan would have to live outside of this function.)

Reply:
Oh yeah all this stuff should get pulled out of here :)